File: java/src/main/java/net/snowflake/floe/FloePurpose.java
package net.snowflake.floe;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;

interface FloePurpose {
  byte[] generate();
}

class HeaderTagFloePurpose implements FloePurpose {
  private static final byte[] bytes = "HEADER_TAG:".getBytes(StandardCharsets.UTF_8);

  static final HeaderTagFloePurpose INSTANCE = new HeaderTagFloePurpose();

  private HeaderTagFloePurpose() {}

  @Override
  public byte[] generate() {
    return bytes;
  }
}

class DekTagFloePurpose implements FloePurpose {
  private static final byte[] prefix = "DEK:".getBytes(StandardCharsets.UTF_8);

  private final byte[] bytes;

  DekTagFloePurpose(long segmentCount) {
    ByteBuffer buffer = ByteBuffer.allocate(prefix.length + 8 /*size of long*/);
    buffer.put(prefix);
    buffer.putLong(segmentCount);
    this.bytes = buffer.array();
  }

  @Override
  public byte[] generate() {
    return bytes;
  }
}

class MessageKeyPurpose implements FloePurpose {
  private static final byte[] bytes = "MESSAGE_KEY:".getBytes(StandardCharsets.UTF_8);

  static final MessageKeyPurpose INSTANCE = new MessageKeyPurpose();

  @Override
  public byte[] generate() {
    return bytes;
  }
}

File: cpp/src/FloePurpose.cpp
#include "FloePurpose.hpp"
#include <algorithm>

namespace floe {

const HeaderTagFloePurpose& HeaderTagFloePurpose::getInstance() {
    static HeaderTagFloePurpose instance;
    return instance;
}

std::vector<uint8_t> HeaderTagFloePurpose::generate() const {
    return {PREFIX.begin(), PREFIX.end()};
}

DekTagFloePurpose::DekTagFloePurpose(const uint64_t segmentCount) {
    bytes_.resize(PREFIX.size() + 8);
    std::ranges::copy(PREFIX, bytes_.begin());
    const uint64_t segmentCountBE = __builtin_bswap64(segmentCount);
    const auto* countBytes = reinterpret_cast<const uint8_t*>(&segmentCountBE);
    std::copy_n(countBytes, 8, bytes_.data() + PREFIX.size());
}

std::vector<uint8_t> DekTagFloePurpose::generate() const {
    return bytes_;
}

const MessageKeyPurpose& MessageKeyPurpose::getInstance() {
    static MessageKeyPurpose instance;
    return instance;
}

std::vector<uint8_t> MessageKeyPurpose::generate() const {
    return {PREFIX.begin(), PREFIX.end()};
}

}
